### PR TITLE
[NP-1958] sequence crunch steps

### PR DESCRIPTION
**Pull request by KernelDeimos**
This PR resolves an anti-pattern in CrunchController that was preventing us from changing the order of Crunch-related screens, re-using them in different situations, etc.

This PR looks large, mostly because a lot of code was rearranged.

### CODE REVIEW

File: src/foam/util/async/Sequence.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+foam.CLASS({
```
Comment (jlhughes) [inline src/foam/util/async/Sequence.js:7]: license header.

File: src/foam/u2/crunch/wizardflow/CapabilityAdaptAgent.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    {
+      name: 'rootCapability',
+      class: 'FObjectProperty',
+      of: 'foam.nanos.crunch.Capability'
+    }
+  ],
+
+  methods: [
+    async function execute() {
+      console.log('umm...', this.unadaptedRootCapability);
```
Comment (tharmaman) [inline src/foam/u2/crunch/wizardflow/CapabilityAdaptAgent.js]: left console.log

File: src/foam/u2/crunch/wizardflow/CapabilityAdaptAgent.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      if ( typeof this.unadaptedRootCapability == 'string' ) {
+        this.rootCapability =
+          await this.capabilityDAO.find(this.unadaptedRootCapability);
+        console.log(this.rootCapability)
```
Comment (tharmaman) [inline src/foam/u2/crunch/wizardflow/CapabilityAdaptAgent.js]: log

File: src/foam/u2/crunch/wizardflow/TestAgent.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+  imports: [
+    'capabilities',
+    'wizardlets'
+  ],
+
+  methods: [
+    // If Property expressions ever unwrap promises this method can be blank.
+    function execute() {
+      console.log('caps', this.capabilities);
```
Comment (tharmaman) [inline src/foam/u2/crunch/wizardflow/TestAgent.js]: logs

File: src/foam/u2/crunch/wizardflow/RequirementsPreviewAgent.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+              capabilityId: this.rootCapability.id,
+              onClose: (x, isContinueAction) => {
+                resolve(isContinueAction);
+              }
+            })
+          ).end();
+        }).then(isContinueAction => {
+          if ( ! isContinueAction ) {
+            this.endSequence();
+            return;
```
Comment (tharmaman) [inline src/foam/u2/crunch/wizardflow/RequirementsPreviewAgent.js]: So execute  is returning null in  this case?

File: src/foam/u2/crunch/wizardflow/CapabilityAdaptAgent.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  properties: [
+    {
+      name: 'rootCapability',
+      class: 'FObjectProperty',
+      of: 'foam.nanos.crunch.Capability'
+    }
+  ],
+
+  methods: [
+    async function execute() {
```
Comment (tharmaman) [inline src/foam/u2/crunch/wizardflow/CapabilityAdaptAgent.js:32]: do problems arise  with this being an async/await function?

File: src/foam/u2/crunch/CrunchController.js [hunk 7]
```diff
@@ -66,183 +63,19 @@ foam.CLASS({
...
+      }))
+        .add(this.CapabilityAdaptAgent)
+        .add(this.LoadCapabilitiesAgent)
+        .add(this.CheckPendingAgent)
+        .add(this.CreateWizardletsAgent)
+        .add(this.RequirementsPreviewAgent)
+        .add(this.StepWizardAgent)
+        .add(this.PutFinalJunctionsAgent)
+        .add(this.TestAgent)
+        ;
```
Comment (tharmaman) [inline src/foam/u2/crunch/CrunchController.js:78]: weird semi colon

**Review by tharmaman — APPROVED** [review]
Just some few questions but no crucial changes

File: src/foam/u2/crunch/wizardflow/CapabilityAdaptAgent.js (hunk 6 above)
Comment (KernelDeimos) [reply src/foam/u2/crunch/wizardflow/CapabilityAdaptAgent.js:32]: Not that I'm aware of. `async` only affects what you can do inside the function, and also means it always returns a promise.

File: src/foam/u2/crunch/CrunchController.js (hunk 7 above)
Comment (KernelDeimos) [reply src/foam/u2/crunch/CrunchController.js:78]: It prevents merge conflicts if two people edit this list at the same time

File: src/foam/u2/crunch/wizardflow/TestAgent.js (hunk 4 above)
Comment (KernelDeimos) [reply src/foam/u2/crunch/wizardflow/TestAgent.js]: `TestAgent` is just for displaying these console logs. I'd like to keep it here but we could comment out TestAgent in the sequence

File: src/foam/u2/crunch/wizardflow/RequirementsPreviewAgent.js (hunk 5 above)
Comment (KernelDeimos) [reply src/foam/u2/crunch/wizardflow/RequirementsPreviewAgent.js]: Yeah and that's not good because Sequence expects it to be async; fixed it

**Pull request merged**